File: obdiggio/obdiggio/ble/mock_transport.py
"""Simulatore ELM327 come :class:`Transport`.

Permette di sviluppare e testare l'intera applicazione (protocollo + UI) su
desktop, senza adattatore e senza veicolo. Simula un motore al minimo con
qualche valore che oscilla, risponde ai comandi AT e restituisce due DTC di
esempio.
"""

from __future__ import annotations

import math
import random
import time

from ..obd.transport import Transport


class MockELM327Transport(Transport):
    """Adattatore ELM327 simulato.

    Alla ``write`` di un comando produce immediatamente nel buffer la risposta
    corrispondente, terminata dal prompt ``>``.
    """

    def __init__(self, dtcs: bool = True) -> None:
        super().__init__()
        self._open = False
        self._t0 = time.monotonic()
        self._has_dtcs = dtcs

    @property
    def is_connected(self) -> bool:
        return self._open

    def open(self) -> None:
        self._open = True

    def close(self) -> None:
        self._open = False

    # --- Simulazione ------------------------------------------------------

    def write(self, data: bytes) -> None:
        if not self._open:
            raise RuntimeError("Trasporto simulato non aperto")
        command = data.decode("ascii", errors="ignore").strip().upper()
        response = self._respond(command)
        # Emetti la risposta come farebbe un vero device (con prompt finale).
        self._feed((response + "\r\r>").encode("ascii"))

    def _respond(self, command: str) -> str:
        if command.startswith("AT"):
            return self._respond_at(command)
        if command.startswith("01"):
            return self._respond_mode01(command)
        if command == "03":
            return self._respond_dtcs()
        if command == "04":
            return "44"
        return "NO DATA"

    def _respond_at(self, command: str) -> str:
        if command == "ATZ":
            return "ELM327 v1.5"
        if command == "ATRV":
            return f"{12.2 + random.uniform(-0.2, 0.4):.1f}V"
        return "OK"

    def _elapsed(self) -> float:
        return time.monotonic() - self._t0

    def _respond_mode01(self, command: str) -> str:
        try:
            pid = int(command[2:4], 16)
        except (ValueError, IndexError):
            return "NO DATA"
        t = self._elapsed()

        def frame(*data_bytes: int) -> str:
            parts = ["41", f"{pid:02X}"] + [f"{b:02X}" for b in data_bytes]
            return " ".join(parts)

        if pid == 0x04:   # carico motore
            return frame(int(30 + 20 * (0.5 + 0.5 * math.sin(t))))
        if pid == 0x05:   # temp refrigerante -> ~90°C (offset +40)
            return frame(min(215, 90 + 40 + int(3 * math.sin(t / 5))))
        if pid == 0x0C:   # RPM ~ 800 al minimo
            rpm = 800 + 300 * (0.5 + 0.5 * math.sin(t / 2))
            raw = int(rpm * 4)
            return frame((raw >> 8) & 0xFF, raw & 0xFF)
        if pid == 0x0D:   # velocita'
            return frame(0)
        if pid == 0x0F:   # temp aria aspirata
            return frame(30 + 40)
        if pid == 0x10:   # MAF
            raw = int((2.0 + 0.5 * math.sin(t)) * 100)
            return frame((raw >> 8) & 0xFF, raw & 0xFF)
        if pid == 0x11:   # farfalla
            return frame(int(255 * 0.15))
        if pid == 0x2F:   # livello carburante
            return frame(int(255 * 0.62))
        if pid == 0x42:   # tensione modulo
            raw = int(12300 + random.uniform(-100, 200))
            return frame((raw >> 8) & 0xFF, raw & 0xFF)
        if pid == 0x46:   # temp ambiente
            return frame(22 + 40)
        # PID non simulato.
        return "NO DATA"

    def _respond_dtcs(self) -> str:
        if not self._has_dtcs:
            return "43 00 00 00"
        # Mode 03: "43" seguito dalle coppie DTC. P0133 (01 33) e P0420 (04 20).
        return "43 01 33 04 20"

File: obdiggio-android/core/src/main/kotlin/org/diggio/obdiggio/core/obd/MockTransport.kt
package org.diggio.obdiggio.core.obd

import kotlin.math.sin
import kotlin.random.Random

/**
 * Simulatore ELM327 come [Transport]: permette di sviluppare e testare
 * l'intera app (protocollo + UI) senza adattatore né veicolo. Alla scrittura di
 * un comando produce subito nel buffer la risposta corrispondente col prompt `>`.
 */
class MockTransport(private val hasDtcs: Boolean = true) : Transport() {

    private var opened = false
    private val t0 = System.nanoTime()

    override val isConnected: Boolean get() = opened

    override fun open() { opened = true }
    override fun close() { opened = false }

    override fun write(data: ByteArray) {
        check(opened) { "Trasporto simulato non aperto" }
        val command = String(data, Charsets.US_ASCII).trim().uppercase()
        feed((respond(command) + "\r\r>").toByteArray(Charsets.US_ASCII))
    }

    private fun elapsed(): Double = (System.nanoTime() - t0) / 1e9

    private fun respond(command: String): String = when {
        command.startsWith("AT") -> respondAt(command)
        // UDS: ReadDTCInformation (0x19/0x02) e ClearDiagnosticInformation (0x14).
        command.startsWith("1902") -> if (hasDtcs)
            "59 02 FF 20 15 11 08 04 01 00 08" else "59 02 FF"
        command.startsWith("14") -> "54"
        command.startsWith("02") -> respondMode02(command)
        command.startsWith("01") -> respondMode01(command)
        // Formato CAN (ISO 15765-4): "43 NN <coppie>", NN = numero di DTC.
        command == "03" -> if (hasDtcs) "43 02 01 33 04 20" else "43 00 00 00 00 00 00"
        command == "04" -> "44"
        else -> "NO DATA"
    }

    private fun respondAt(command: String): String = when (command) {
        "ATZ" -> "ELM327 v1.5"
        "ATRV" -> "%.1fV".format(12.2 + Random.nextDouble(-0.2, 0.4))
        else -> "OK"
    }

    // Freeze frame simulato (Mode 02): snapshot "al momento del guasto".
    private fun respondMode02(command: String): String {
        val pid = command.substring(2, 4).toIntOrNull(16) ?: return "NO DATA"
        fun frame(vararg d: Int) =
            (listOf("42", "%02X".format(pid), "00") + d.map { "%02X".format(it) }).joinToString(" ")
        return when (pid) {
            0x02 -> "42 02 00 01 04"                       // DTC che ha congelato: P0104
            0x04 -> frame((255 * 0.78).toInt())            // carico 78%
            0x05 -> frame(92 + 40)                         // refrigerante 92°C
            0x0B -> frame(158)                             // MAP 158 kPa
            0x0C -> { val r = 1850 * 4; frame((r shr 8) and 0xFF, r and 0xFF) } // 1850 rpm
            0x0D -> frame(62)                              // 62 km/h
            0x0E -> frame(128)                             // anticipo 0°
            0x0F -> frame(35 + 40)                         // aria aspirata 35°C
            0x10 -> { val r = 1500; frame((r shr 8) and 0xFF, r and 0xFF) }     // MAF 15 g/s
            0x11 -> frame((255 * 0.42).toInt())            // farfalla 42%
            0x1F -> { val r = 320; frame((r shr 8) and 0xFF, r and 0xFF) }      // 320 s dall'avvio
            0x33 -> frame(101)                             // barometrica
            else -> "NO DATA"
        }
    }

    private fun respondMode01(command: String): String {
        val pid = command.substring(2, 4).toIntOrNull(16) ?: return "NO DATA"
        val t = elapsed()
        fun frame(vararg dataBytes: Int): String =
            (listOf("41", "%02X".format(pid)) + dataBytes.map { "%02X".format(it) }).joinToString(" ")
        return when (pid) {
            // Maschere dei PID supportati (per il rilevamento automatico).
            0x00 -> "41 00 18 3B 80 01"
            0x20 -> "41 20 00 02 20 01"
            0x40 -> "41 40 44 80 00 00"
            0x04 -> frame((30 + 20 * (0.5 + 0.5 * sin(t))).toInt())
            0x05 -> frame(minOf(215, 90 + 40 + (3 * sin(t / 5)).toInt()))
            0x0B -> frame((104 + 18 * (0.5 + 0.5 * sin(t))).toInt()) // MAP (kPa)
            0x0C -> {
                val raw = ((800 + 300 * (0.5 + 0.5 * sin(t / 2))) * 4).toInt()
                frame((raw shr 8) and 0xFF, raw and 0xFF)
            }
            0x0D -> frame(0)
            0x0F -> frame(30 + 40)
            0x10 -> {
                val raw = ((2.0 + 0.5 * sin(t)) * 100).toInt()
                frame((raw shr 8) and 0xFF, raw and 0xFF)
            }
            0x11 -> frame((255 * 0.15).toInt())
            0x2F -> frame((255 * 0.62).toInt())
            0x33 -> frame(101)                                       // barometrica (kPa)
            0x42 -> {
                val raw = (12300 + Random.nextDouble(-100.0, 200.0)).toInt()
                frame((raw shr 8) and 0xFF, raw and 0xFF)
            }
            0x46 -> frame(22 + 40)
            0x49 -> frame((255 * 0.12).toInt())                      // pedale acceleratore
            else -> "NO DATA"
        }
    }
}
